### PR TITLE
Set custom cursor theme when RadioButton and CheckBox are disabled

### PR DESCRIPTION
**Pull request by hagatopaxi**
#### What does this PR do?

When a CheckBox or a RadioButton is disabled, it's possible to set a custom cursor, applied on the container components (the cursor is also applied on the label text). 

#### Where should the reviewer start?

src/js/components/CheckBox/StyledCheckBox.js
src/js/components/RadioButton/StyledRadioButton.js

#### What testing has been done on this PR?

Manual and automated tests

#### How should this be manually tested?

Override the default theme with `<Grommet theme={{ checkBox: { disabled: { cursor: "not-allowed" } }, radioButton: { disabled: { cursor: "not-allowed" } } }}>` 

#### Do Jest tests follow these best practices?

- [ x] `screen` is used for querying.
- [ x] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ x] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

#### What are the relevant issues?

[https://github.com/grommet/grommet/issues/6862](https://github.com/grommet/grommet/issues/6862)

#### Screenshots (if appropriate)

image
image

#### Do the grommet docs need to be updated?

Yes

#### Should this PR be mentioned in the release notes?

Yes

#### Is this change backwards compatible or is it a breaking change?

No

### CODE REVIEW

**Comment by jcfilben** [comment]
Hey @LegerG sorry for the delay reviewing this, hoping to get a chance to take a closer look at this early next week. Thanks for your patience!

**Review by jcfilben — COMMENTED** [review]
I'm thinking that we could handle setting the cursor style with a  `global.input.disabled.cursor` prop in the theme instead of needing `checkbox.disabled` or `radioButton.disabled`. This would keep the theme structure a little simpler as well as allow for setting the stying of the disabled cursor across all input components.

**Comment by LegerG** [comment]
Thank you for your response, I have updated the PR.
I will update other inputs in the next PR.

**Comment by jcfilben** [comment]
Hey @LegerG thanks for updating the PR! I'm thinking support should be added for `global.input.disabled.cursor` to all the input components within one PR. I know you mentioned tackling the rest of this work in a subsequent PR but I think ideally we don't want to have a gap of time where we have added this global theme prop but it is only working for some of the input components.

**Comment by LegerG** [comment]
Hello, I have almost done the code but I have some points I need to discuss with you

1. RangeSelector and DateInput don't have disabled props.
2. Anchor and Button are not considered as inputs in the documentation but they are included in some components like SelectMultiple or FileInput. So there are some parts of the component where the cursor is different (see the screen below).
3. MaskedInput documentation misses the disabled props, but I have implemented the feature anyway, because the props are in the code.
4. DefaultSelectTextInput in the Select components has a comment about the disabled props. I have not implemented the feature for this one. I don't know what is the best to do. [](https://github.com/grommet/grommet/blob/master/src/js/components/Select/DefaultSelectTextInput.js#L7)
5. SelectMultiple has a bug with the showSelectedInline props and disabled. The showed selected items are not disabled. I have not fixed this feature

Different cursor inside the component
image
image


SelectMultiple issue when disabled
image
image

If you need more details about one of these points, don't hesitate. :)

**Comment by taysea** [comment]
Hey there, thanks for your contribution so far and your willingness to incorporate feedback! After a bit more discussion within our team, given [your comment](https://github.com/grommet/grommet/pull/6867#issuecomment-1696830333), we've decided it would be best to support this from the `theme.global.control.disabled.cursor` object instead of the previous discussion around `theme.global.input`.

I imagine in `disabledStyle`, the `cursor` line could look something like `theme.global.control.disabled.cursor || 'default'`. Would you be willing to make that adjustment?

With this direction, Anchor and Button would also receive this treatment and so would any inputs that support `disabled` props. Regarding documentation of `disabled` prop on grommet docs, our approach has been not to document properties that are native HTML properties.

Thank you for your notes about RangeSelector/DateInput `disabled` support -- we'll file separate tickets to look into those. And I already filed one for the [SelectMultiple issue](https://github.com/grommet/grommet/issues/6945) you mentioned. We can take care of those issues separate from this PR.

**Comment by LegerG** [comment]
I have pushed for a new version following your requirements. It is big PR, so there is a list of all components impacted in the source code (for the others, it is only tests or stories):
* Anchor
* Button
* CheckBox
* FileInput
* RadioButton
* RangeInput
* Select

For now, I haven't written any tests (except for the RadioButton and CheckBox components). I plan to do it in the next commit. Do I need to write tests for all components affected by the update, or only those whose source code has been modified? (for example, only for RadioButton and not for RadioButtonGroup).

Don't hesitate if you need more details 😃

File: src/js/components/Select/stories/Disabled.js [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
-        name="select"
-        placeholder="Select"
-        value={value}
-        options={options}
-        onChange={({ option }) => setValue(option)}
-        disabled
-      />
-    </Box>
-    // </Grommet>
+    <Grommet
```
Comment (taysea) [inline src/js/components/Select/stories/Disabled.js:9]: If this is a custom themed story, it should be under a "custom theme" folder similar to: https://storybook.grommet.io/?path=/story/controls-accordion-custom-themed-custom--custom

File: src/js/components/Select/StyledSelect.js [hunk 2]
```diff
@@ -40,15 +40,15 @@ export const HiddenInput = styled.input`
 export const SelectOption = styled(Button)`
   ${(props) => props.selected && props.textComponent && selectedStyle}
   // applies theme.global.hover.background to the active
-  // option for mouse and keyboard interactions
+    // option for mouse and keyboard interactions
```
Comment (taysea) [inline src/js/components/Select/StyledSelect.js:43]: Seems like some unnecessary space was added here, potentially due to different local linting/prettier set up?

File: src/js/components/Button/StyledButton.js [hunk 3]
```diff
@@ -162,6 +162,7 @@ const activeButtonStyle = (props) => css`
 
 const disabledButtonStyle = (props) => css`
   ${disabledStyle(props.theme.button.disabled.opacity)}
+  ${props.theme.global.input.disabled && props.theme.global.input.disabled}
```
Comment (taysea) [inline src/js/components/Button/StyledButton.js:165]: It seems odd to be applying an `input` styling here on button. I also think the theme object has changed. Can this be removed?

File: src/js/components/Button/stories/Disabled.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+
+import { Box, Button, Grommet } from 'grommet';
+
+export const Disabled = () => (
+  <Grommet
+    theme={{ global: { control: { disabled: { cursor: 'not-allowed' } } } }}
```
Comment (taysea) [inline src/js/components/Button/stories/Disabled.js:7]: Instead of adding a new story, can this theme just be added to this story https://storybook.grommet.io/?path=/story/controls-button-custom-themed-ts-custom--ts-custom?

File: src/js/components/CheckBox/stories/Simple.js [hunk 5]
```diff
@@ -14,10 +14,12 @@ const SimpleCheckBox = ({ checked: checkedProp, ...rest }) => {
 };
 
 export const Simple = () => (
-  <>
+  <Grommet
```
Comment (taysea) [inline src/js/components/CheckBox/stories/Simple.js:17]: No need to modify this story. The concept is sufficiently demonstrated in other stories.

File: src/js/components/FileInput/StyledFileInput.js [hunk 6]
```diff
@@ -11,20 +11,23 @@ const StyledFileInput = styled.input`
   font-size: 0;
   opacity: 0;
   border: none;
-  ${(props) => !props.disabled && 'cursor: pointer;'}
+  cursor: ${(props) =>
+    props.disabled ? props.theme.global.control.disabled.cursor : 'pointer'};
```
Comment (taysea) [inline src/js/components/FileInput/StyledFileInput.js:15]: Should this be?

```suggestion
    props.disabled && props.theme.global.control.disabled.cursor ? props.theme.global.control.disabled.cursor : 'pointer'};
```

**Comment by LegerG** [comment]
Hello, I have applied all your feedbacks, thanks for the review.

File: src/js/components/Button/StyledButton.js [hunk 7]
```diff
@@ -162,6 +162,7 @@ const activeButtonStyle = (props) => css`
 
 const disabledButtonStyle = (props) => css`
   ${disabledStyle(props.theme.button.disabled.opacity)}
+  ${props.theme.global.input.disabled}
```
Comment (taysea) [inline src/js/components/Button/StyledButton.js:165]:
```suggestion
```
This looks like it's lingering from earlier state of PR and can be removed.

File: src/js/components/Button/StyledButton.js [hunk 8]
```diff
@@ -181,6 +182,9 @@ const disabledButtonStyle = (props) => css`
         props.theme,
       )};`)}
   ${props.theme.button.disabled && props.theme.button.disabled.extend}
+  cursor: ${props.disabled
```
Comment (taysea) [inline src/js/components/Button/StyledButton.js:185]: Since this is already in the disabled style function, should this just be:
```
cursor: props.theme.global.control.disabled.cursor || 'default';
```

Also, can you move this above the `extend` in the line above? `extend` should always be the last thing applied.

File: src/js/components/Button/__tests__/Button-kind-test.js [hunk 9]
```diff
@@ -1,843 +0,0 @@
...
-import { axe } from 'jest-axe';
-import { fireEvent, render, screen } from '@testing-library/react';
-import { Add } from 'grommet-icons';
-
-import { hpe } from 'grommet-theme-hpe';
-import { deepMerge } from '../../../utils';
-import { Grommet, Button, Box } from '../..';
-import { buttonKindTheme } from './theme/buttonKindTheme';
-
-describe('Button kind', () => {
```
Comment (taysea) [inline src/js/components/Button/__tests__/Button-kind-test.js:16]: I don't think this file should be deleted.

File: src/js/components/DataTable/__tests__/DataTable-test.tsx [hunk 10]
```diff
@@ -552,7 +552,7 @@ describe('DataTable', () => {
             }
             return (
               <Box>
-                {row.a} : {row.b}{' '}
+                {row.a} :{row.b}
```
Comment (taysea) [inline src/js/components/DataTable/__tests__/DataTable-test.tsx:555]: I don't think this file should be changed by this PR, looks unrelated/potentially a local linting change

File: src/js/components/FileInput/FileInput.js [hunk 11]
```diff
@@ -55,6 +55,8 @@ const Label = styled(Text)`
     props.theme.fileInput &&
     props.theme.fileInput.label &&
     props.theme.fileInput.label.extend};
+  cursor: ${(props) =>
```
Comment (taysea) [inline src/js/components/FileInput/FileInput.js:58]: Similar comment to above, but this should appear before the extend. Also, instead of adding `default` as a fallback, could we simplify the code in all instances as:

```
props.disabled && props.theme.global.controls.disabled.cursor ? `cursor: ${props.theme.global.controls.disabled.cursor};` : '';
```

Comment (taysea) [reply src/js/components/FileInput/FileInput.js:58]: This should simplify the CSS to not add cursor: default in as many places.

File: src/js/components/FileInput/stories/Disabled.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+
+import { Box, FileInput, Grommet } from 'grommet';
+
+export const Disabled = () => (
+  <Grommet
```
Comment (taysea) [inline src/js/components/FileInput/stories/Disabled.js:6]: Since this has a custom theme, can you place it under a CustomThemed folder like your Anchor story?

File: src/js/components/FileInput/stories/Multiple.js [hunk 13]
```diff
@@ -1,27 +1,41 @@
-import React from 'react';
+import React, { useState } from 'react';
```
Comment (taysea) [inline src/js/components/FileInput/stories/Multiple.js:1]: Since the "Disabled" story already captures this behavior, can we leave this story untouched?

File: src/js/components/Grommet/__tests__/__snapshots__/Grommet-test.js.snap [hunk 14]
```diff
@@ -202,7 +202,7 @@ exports[`Grommet hpe theme 1`] = `
   font-size: 18px;
   line-height: 24px;
   background-color: #FFFFFF;
-  color: #555555;
+  color: #6F6F6F;
```
Comment (taysea) [inline src/js/components/Grommet/__tests__/__snapshots__/Grommet-test.js.snap:205]: Looks like this is out of date with master.

File: src/js/components/MaskedInput/stories/Disabled.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+import { Grommet } from '../../Grommet';
+
+const daysInMonth = (month) => new Date(2019, month, 0).getDate();
+
+export const Disabled = () => {
+  const [value, setValue] = React.useState('');
+
+  return (
+    <Grommet
+      theme={{ global: { control: { disabled: { cursor: 'not-allowed' } } } }}
```
Comment (taysea) [inline src/js/components/MaskedInput/stories/Disabled.js:12]: Same comment as above about putting this in a CustomThemed folder.

File: src/js/components/RadioButton/stories/Disabled.js [hunk 16]
```diff
@@ -1,17 +1,21 @@
...
-  <Box align="center" pad="large" gap="large">
-    <RadioButton
-      label="option 1"
-      name="name"
-      value="option 1"
-      checked
-      disabled
-    />
-  </Box>
+  <Grommet
```
Comment (taysea) [inline src/js/components/RadioButton/stories/Disabled.js:6]: I think there are enough stories demoing this behavior, can we leave this story as it was?

File: src/js/components/RangeInput/stories/Disabled.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { Box, Grommet, RangeInput } from 'grommet';
+
+export const Disabled = () => {
+  const [value, setValue] = React.useState(5);
+
+  const onChange = (event) => setValue(event.target.value);
+
+  return (
+    <Grommet
```
Comment (taysea) [inline src/js/components/RangeInput/stories/Disabled.js:10]: I think there are enough stories demoing this behavior, can we leave this story as it was?

File: src/js/components/Select/__tests__/__snapshots__/Select-test.js.snap [hunk 18]
```diff
@@ -5905,6 +5909,8 @@ exports[`Select keyboard navigation timeout 1`] = `
   border-radius: 4px;
   outline: none;
   border: none;
+  opacity: 0.3;
```
Comment (taysea) [inline src/js/components/Select/__tests__/__snapshots__/Select-test.js.snap:5912]: It seems suspicious that this styling is being added here even though nothing changed in select-test.

**Comment by jcfilben** [comment]
Closing due to inactivity, feel free to re-open if this feature is still desired
